ml/src/test/scala/org/apache/spark/ml/h2o/features/H2OTargetEncoderTest.scala

```diff
@@ -0,0 +1,88 @@
+import org.junit.runner.RunWith
+import org.scalatest.FunSuite
+import org.scalatest.junit.JUnitRunner
+import water.api.TestUtils
+
+@RunWith(classOf[JUnitRunner])
+class H2OTargetEncoderTest extends FunSuite with SharedH2OTestContext {
+
+  override def createSparkContext = new SparkContext("local[*]", "H2OTargetEncoderTest", conf = defaultSparkConf)
+
+  private def loadDataFrameFromCsv(path: String): DataFrame = {
+    spark.read
+      .option("header", "true")
+      .option("inferSchema", "true")
+      .csv(TestUtils.locate(path))
+  }
+
+  private def loadDataFrameFromCsvAsResource(path: String): DataFrame = {
+    val filePath = getClass.getResource(path).getFile
+    spark.read
+      .option("header", "true")
+      .option("inferSchema", "true")
+      .csv(filePath)
+  }
+
+  private def assertDataFramesAreIdentical(expected: DataFrame, produced: DataFrame): Unit = {
+    produced.cache()
+    val numberOfExtraColumnsInExpected = expected.exceptAll(produced).count()
+    val numberOfExtraColumnsInProduced = produced.exceptAll(expected).count()
+    produced.unpersist()
+    assert(
+      numberOfExtraColumnsInExpected == 0 && numberOfExtraColumnsInProduced == 0,
+      s"""The expected data frame contains $numberOfExtraColumnsInExpected rows that are not in the produced data frame.
+         |The produced data frame contains $numberOfExtraColumnsInProduced rows that are not in the expected data frame.
+       """.stripMargin)
+  }
+
+  private lazy val dataset = loadDataFrameFromCsv("smalldata/prostate/prostate.csv")
+  private lazy val Array(trainingDataset, testingDataset) = dataset.randomSplit(Array(0.8, 0.2), 1234L).map(_.cache())
+  private lazy val expectedTestingDataset = loadDataFrameFromCsvAsResource("/target_encoder/testing_dataset_transformed.csv").cache()
+
+  test("The pipeline with a target encoder transform training and testing dataset without an exception") {
+    val targetEncoder = new H2OTargetEncoder()
+      .setInputCols(Array("RACE", "DPROS", "DCAPS"))
+      .setLabelCol("CAPSULE")
+    val gbm = new H2OGBM().setLabelCol("CAPSULE")
+
+    val pipeline = new Pipeline().setStages(Array(targetEncoder, gbm))
+
+    val model = pipeline.fit(trainingDataset)
+    model.transform(testingDataset)
+  }
+
+  test("The target encoder is able to transform dataset after it's saved and loaded") {
+    val targetEncoder = new H2OTargetEncoder()
+      .setInputCols(Array("RACE", "DPROS", "DCAPS"))
+      .setLabelCol("CAPSULE")
+    val pipeline = new Pipeline().setStages(Array(targetEncoder))
+
+    val model = pipeline.fit(trainingDataset)
+    val path = "build/ml/targetEncoder_save_load"
+    model.write.overwrite().save(path)
+    val loadedModel = PipelineModel.load(path)
+    val transformedTestingDataset = model.transform(testingDataset)
+
+    assertDataFramesAreIdentical(expectedTestingDataset, transformedTestingDataset)
+  }
+
+  test("The target encoder doesn't apply noise on the testing dataset") {
```

Comment:
@mn-mikke Having a hard time to understand this test. Could you please help me with this. So... we call `fit` on our pipeline...it means that in H2OTargetEncoderModel we are in `inTrainingMode` mode, right? And we have following logic `val noise = Option(getNoise()).getOrElse(H2OTargetEncoderNoiseSettings(amount = 0.0))` and it feels like we should apply noise as we set it to 0.5.  And it is not `H2OTargetEncoderMojoModel`. What am I missing?

Reply:
`inTrainingMode` is valid only for [fitting](https://github.com/h2oai/sparkling-water/pull/1192/files#diff-7908001f8c9b38a0a2354f5996b51c6aR83). At that time ,noise is applied according to the parameters. When testing dataset is [transformed](https://github.com/h2oai/sparkling-water/pull/1192/files#diff-7908001f8c9b38a0a2354f5996b51c6aR84) we are not `inTrainingMode` anymore and noise won't be applied.

 Does it make sense?

Reply:
@mn-mikke the question is how do we switch from fitting/`inTrainingMode`  to not `inTrainingMode`?  Because it feels like StackTraceElement which satisfies `inTrainingMode` conditions could be still there.

Reply:
> Because it feels like StackTraceElement which satisfies inTrainingMode conditions could be still there.

@deil87 Please can you give me an example?